Modules/_testinternalcapi/test_critical_sections.c

```diff
@@ -0,0 +1,204 @@
+{
+    PyObject *a = PyDict_New();
+    assert(a != NULL);
+    PyObject *b = PyDict_New();
+    assert(b != NULL);
+
+    // Locking an object recursively with this API should not deadlock.
+    Py_BEGIN_CRITICAL_SECTION(a);
+    lock_unlock_object(a, 10);
+    assert_nogil(PyMutex_IsLocked(&a->ob_mutex));
+    Py_END_CRITICAL_SECTION;
+
+    // Same test but with two objects.
+    Py_BEGIN_CRITICAL_SECTION2(b, a);
+    lock_unlock_two_objects(a, b, 10);
+    assert_nogil(PyMutex_IsLocked(&a->ob_mutex));
+    assert_nogil(PyMutex_IsLocked(&b->ob_mutex));
+    Py_END_CRITICAL_SECTION2;
+
+    Py_DECREF(b);
+    Py_DECREF(a);
+    Py_RETURN_NONE;
+}
+
+// Test that a critical section is suspened by a Py_BEGIN_ALLOW_THREADS and
+// resumed by a Py_END_ALLOW_THREADS.
+static PyObject *
+test_critical_sections_suspend(PyObject *self, PyObject *Py_UNUSED(args))
+{
+    PyObject *a = PyDict_New();
+    assert(a != NULL);
+
+    Py_BEGIN_CRITICAL_SECTION(a);
+    assert_nogil(PyMutex_IsLocked(&a->ob_mutex));
+
+    // Py_BEGIN_ALLOW_THREADS should suspend the active critical section
+    Py_BEGIN_ALLOW_THREADS
+    assert_nogil(!PyMutex_IsLocked(&a->ob_mutex));
+    Py_END_ALLOW_THREADS;
+
+    // After Py_END_ALLOW_THREADS the critical section should be resumed.
+    assert_nogil(PyMutex_IsLocked(&a->ob_mutex));
+    Py_END_CRITICAL_SECTION;
+
+    Py_DECREF(a);
+    Py_RETURN_NONE;
+}
+
+struct test_data {
+    PyObject *obj1;
+    PyObject *obj2;
+    PyObject *obj3;
+    Py_ssize_t countdown;
+    PyEvent done_event;
+};
+
+static void
+thread_critical_sections(void *arg)
+{
+    const Py_ssize_t NUM_ITERS = 200;
+    struct test_data *test_data = arg;
+    PyGILState_STATE gil = PyGILState_Ensure();
+
+    for (Py_ssize_t i = 0; i < NUM_ITERS; i++) {
+        Py_BEGIN_CRITICAL_SECTION(test_data->obj1);
+        Py_END_CRITICAL_SECTION;
+
+        Py_BEGIN_CRITICAL_SECTION(test_data->obj2);
+        lock_unlock_object(test_data->obj1, 1);
+        Py_END_CRITICAL_SECTION;
+
+        Py_BEGIN_CRITICAL_SECTION2(test_data->obj3, test_data->obj1);
+        lock_unlock_object(test_data->obj2, 2);
+        Py_END_CRITICAL_SECTION2;
+
+        Py_BEGIN_CRITICAL_SECTION(test_data->obj3);
+        Py_BEGIN_ALLOW_THREADS
+        Py_END_ALLOW_THREADS
+        Py_END_CRITICAL_SECTION;
+    }
+
+    PyGILState_Release(gil);
+    if (_Py_atomic_add_ssize(&test_data->countdown, -1) == 1) {
+        // last thread to finish sets done_event
+        _PyEvent_Notify(&test_data->done_event);
+    }
+}
+
+static PyObject *
+test_critical_sections_threads(PyObject *self, PyObject *Py_UNUSED(args))
+{
+    const Py_ssize_t NUM_THREADS = 4;
+    struct test_data test_data = {
+        .obj1 = PyDict_New(),
+        .obj2 = PyDict_New(),
+        .obj3 = PyDict_New(),
+        .countdown = NUM_THREADS,
+    };
+    assert(test_data.obj1 != NULL);
+    assert(test_data.obj2 != NULL);
+    assert(test_data.obj3 != NULL);
```

Comment:
Just a drive-by comment, I would much prefer to have conditional returns from native test functions than asserts. When these fail, they can block test progress (on Windows it pops up a modal dialog by default), or may trigger error reporting or a debugger. It also shows up as a test crash, rather than a failure.

Perhaps a new macro is needed to print location information to the console and then fail?

Reply:
I agree that the failure case is unpleasant, but I don't think trying to route proper error returns in the C API tests is worthwhile.  To the extent that there are bugs in the tested code, they are likely to show up as crashes, deadlocks, or other unpleasant behavior even if we make the asserts more nicely behaved (because this is testing C code, not Python). Additionally, it's challenging to route error messages from other threads in C since they don't always have a GIL context and might not be able to create Python error objects.

One consequence of this is that I think we should basically have a zero tolerance for flaky C API tests. If we are going to be bothered by an unpleasant failure, it better signify something important.

If crashes in the C API tests become an actual problem, then I think it would be better to invest in running these tests in a subprocess like we do with [`test_support.script_helper.assert_python_failure`](https://github.com/python/cpython/blob/102685c4c8481ec5d9c132fcf06b46057e815969/Lib/test/support/script_helper.py#L169-L177).

Reply:
Yeah, that's reasonable. I'm mostly annoyed about it because some versions of MSVC have a compiler bug on ARM64 that causes some of the interlocked tests to fail, which means the entire test run is basically useless until all the machines get updated (as far as I can tell, the code is only ever tested and never actually used, so it doesn't need fixing for our releases - there's nothing we can do about it on our side anyway).